src/lustre/lustreTypeChecker.ml

```diff
@@ -0,0 +1,1133 @@
+  | LA.RestartEvery (pos, node, args, cond) ->
+     R.bind(check_type_expr ctx cond (LA.Bool pos)) (fun _ ->
+         infer_type_expr ctx (LA.Call (pos, node, args)))
+                                
+  (* Temporal operators *)
+  | LA.Pre (pos, e) -> infer_type_expr ctx e
+  | LA.Last (pos, i) -> infer_type_expr ctx (LA.Ident (pos, i))
+  | LA.Fby (pos, e1, _, e2) ->
+     R.bind (infer_type_expr ctx e1) (fun ty1 ->
+         R.bind(infer_type_expr ctx e2)(fun ty2 ->
+             R.bind(eq_lustre_type ctx ty1 ty2)(fun is_eq ->
+                 if is_eq
+                 then R.ok ty1
+                 else type_error pos ("Both the expressions in Fby should be of the same type."
+                                     ^ "Found types " ^ string_of_tc_type ty1
+                                     ^ " and " ^ string_of_tc_type ty2))))
+  | LA.Arrow (pos, e1, e2) ->
+     R.bind (infer_type_expr ctx e1) (fun ty1 ->
+         R.bind (infer_type_expr ctx e2) (fun ty2 ->
+             R.bind((eq_lustre_type ctx ty1 ty2)) (fun is_eq -> 
+                 if is_eq then R.ok ty1 else
+                   type_error pos
+                     ("Arrow types do not match "
+                      ^ string_of_tc_type ty1
+                      ^ " and " ^ string_of_tc_type ty2)))) 
+
+  (* Node calls *)
+  | LA.Call (pos, i, arg_exprs) ->
+     Log.log L_debug "Inferring type for node call %a" LA.pp_print_ident i  
+    ; let rec infer_type_node_args: tc_context -> LA.expr list -> tc_type tc_result
+        = fun ctx args ->
+        R.bind (R.seq (List.map (infer_type_expr ctx) args)) (fun arg_tys ->
+            R.ok (LA.TupleType (pos, arg_tys))) in
+      (match (lookup_ty ctx i) with 
+       | TArr (_, exp_arg_tys, exp_ret_tys) ->
+          R.bind (infer_type_node_args ctx arg_exprs) (fun given_arg_tys ->
+              R.bind (eq_lustre_type ctx exp_arg_tys given_arg_tys) (fun is_arg_ty_ok ->
+                  if is_arg_ty_ok
+                  then R.ok exp_ret_tys
+                  else type_error pos ("Node arguments expected to have type "
+                                       ^ string_of_tc_type exp_arg_tys
+                                       ^ " but found type "
+                                       ^ string_of_tc_type given_arg_tys)))
+       | ty -> type_error pos
+                 ("Expected node type to be a function type, but found type"
+                  ^ string_of_tc_type ty)) 
+      
+  | LA.CallParam _ -> Lib.todo __LOC__
+(** Infer the type of a [LA.expr] with the types of free variables given in [tc_context] *)
+
+and check_type_expr: tc_context -> LA.expr -> tc_type -> unit tc_result
+  = fun ctx expr exp_ty ->
+  match expr with
+  (* Identifiers *)
+  | Ident (pos, i) as ident ->
+     R.bind (infer_type_expr ctx ident) (fun ty ->
+         R.bind (eq_lustre_type ctx ty exp_ty) (fun is_eq ->
+             if is_eq
+             then R.ok ()
+             else type_error pos ("Variable " ^ i
```

Comment:
A variable is not the only identifier you can find here. You can also find the name of an enumeration type value. For instance, replace `blue` with `yellow` in `tests/lustre/typechecking/shouldPass/test_enums.lus` line 8, and you will get:

> <Error> line 8 col. 13: Failure cause: Type error: Variable yellow does not match expected type color1 with infered type color2

Change error message to cover all cases.

Reply:
I have changed this to `Identifer` for now. 
What is the term that users of the system would be most comfortable with?

Reply:
I think identifier is fine.